javaslang/src/main/java/javaslang/collection/Collections.java

```diff
@@ -289,29 +289,32 @@ int size() {
         }
     }
 
-    static <T, U extends Iterable<? extends Iterable<T>>> U transpose(U source, Function<Iterable<?>, Iterable<?>> mapper) {
-        if (source.iterator().hasNext()) {
-            return transposeIterables(source, mapper);
-        } else {
+    static <T, U extends Seq<T>, V extends Seq<U>> V transpose(V source, Function<Iterable<U>, V> outerMapper, Function<Iterable<T>, U> innerMapper) {
+        if (source.isEmpty() || (source.length() == 1 && source.head().length() <= 1)) {
             return source;
+        } else {
+            return transposeIterables(source, outerMapper, innerMapper);
         }
     }
 
-    @SuppressWarnings("unchecked")
-    private static <T, U extends Iterable<? extends Iterable<T>>> U transposeIterables(U source, Function<Iterable<?>, Iterable<?>> mapper) {
-        final ArrayList<ArrayList<T>> results = new ArrayList<>();
+    private static <T, U extends Seq<T>, V extends Seq<U>> V transposeIterables(V source, Function<Iterable<U>, V> outerMapper, Function<Iterable<T>, U> innerMapper) {
```

Comment:
Yes, sounds good 😊 
I will push an update that includes

* CharSeq.transpose(Seq&lt;CharSeq>>)
* better names (rowFactory, columnFactory)
* the matrix checks (throwing IllegalArgumentException is ok)